firedrake/adjoint_utils/function.py

```diff
@@ -62,15 +62,15 @@ def wrapper(self, *args, **kwargs):
                 output = subfunctions(self, *args, **kwargs)
 
             if annotate:
-                output = tuple(firedrake.Function(output[i].function_space(),
-                                                  output[i],
-                                                  block_class=SubfunctionBlock,
-                                                  _ad_floating_active=True,
-                                                  _ad_args=[self, i],
-                                                  _ad_output_args=[i],
-                                                  output_block_class=FunctionMergeBlock,
-                                                  _ad_outputs=[self],
-                                                  ad_block_tag=ad_block_tag)
+                output = tuple(type(self)(output[i].function_space(),
+                                          output[i],
+                                          block_class=SubfunctionBlock,
+                                          _ad_floating_active=True,
+                                          _ad_args=[self, i],
+                                          _ad_output_args=[i],
+                                          output_block_class=FunctionMergeBlock,
```

Comment:
This relies on the `FunctionMergeBlock`, which I suspect is not fully functional for cofunctions. For instance, you will probably need to also update [this line](https://github.com/firedrakeproject/firedrake/blob/1bbd9dfa3b9a7dc7e501cc094b93067e89c6448c/firedrake/adjoint_utils/blocks/function.py#L263), which will probably fail when computing the TLM since `Function.assign` and `Cofunction.assign` have a different procedure.

You probably just need to fix the